lib/SILGen/SILGenStmt.cpp

```diff
@@ -908,7 +922,15 @@ void StmtEmitter::visitForEachStmt(ForEachStmt *S) {
   // to hold the results.  This will be initialized on every entry into the loop
   // header and consumed by the loop body. On loop exit, the terminating value
   // will be in the buffer.
-  auto optTy = S->getIteratorNext()->getType()->getCanonicalType();
+  CanType optTy;
+  if (S->getConvertElementExpr()) {
+    optTy = S->getConvertElementExpr()->getType()->getCanonicalType();
+  } else {
+    optTy = cast<FuncDecl>(S->getGeneratorNext().getDecl())
```

Comment:
It might be cleaner to pull this out of the conformance instead -- it's just OptionalType::get() wrapping the element associated type. You can use `getTypeWitnessByName()`